### PR TITLE
Unit Tests and Enhancements for Licensing Components

### PR DESCRIPTION
**Pull request by kingster-will**
This PR includes unit tests for our licensing components. Comprehensive unit tests have been added for the LicensingModul, PILicenseTemplate, LicenseToken, and LicenseRegistory. And addressed TODO's in the codes.

### CODE REVIEW

File: contracts/registries/LicenseRegistry.sol [hunk 1]
```diff
@@ -193,14 +193,23 @@ contract LicenseRegistry is ILicenseRegistry, AccessManagedUpgradeable, UUPSUpgr
...
+        LicenseRegistryStorage storage $ = _getLicenseRegistryStorage();
+        if (_isExpiredNow(ipId)) {
+            revert Errors.LicenseRegistry__IpExpired(ipId);
+        }
+
+        if (_hasIpAttachedLicenseTerms(ipId, licenseTemplate, licenseTermsId)) {
+            revert Errors.LicenseRegistry__LicenseTermsAlreadyAttached(ipId, licenseTemplate, licenseTermsId);
+        }
+
         if (_isDerivativeIp(ipId)) {
```
Comment (Ramarti) [inline contracts/registries/LicenseRegistry.sol:205]: Is this check redundant with the one in line 235? Derivatives will always have license terms attached

**Review by Ramarti — APPROVED** [review]
LGTM

File: contracts/registries/LicenseRegistry.sol [hunk 2]
```diff
@@ -193,14 +193,23 @@ contract LicenseRegistry is ILicenseRegistry, AccessManagedUpgradeable, UUPSUpgr
             revert Errors.LicensingModule__LicenseTermsNotFound(licenseTemplate, licenseTermsId);
         }
 
+        LicenseRegistryStorage storage $ = _getLicenseRegistryStorage();
```
Comment (LeoHChen) [inline contracts/registries/LicenseRegistry.sol:196]: why move this line here?  shouldn't it be at 208?

**Comment by jdubpark** [comment]
Please address issue #41 in the license system

File: test/foundry/integration/e2e/e2e.t.sol
Comment (jdubpark) [inline test/foundry/integration/e2e/e2e.t.sol:1]: Please have a better name for this license-specific integration test file.

File: test/foundry/modules/licensing/LicensingModule.t.sol
Comment (jdubpark) [inline test/foundry/modules/licensing/LicensingModule.t.sol:1]: Per issue #41, there should be a test case where it should revert when registering a derivative IP using two LTs of the same license terms (e.g. NC Social Remix) from the same parent.

Comment (LeoHChen) [reply test/foundry/modules/licensing/LicensingModule.t.sol:1]: Do you plan to fix #41 in a separate PR? @kingster-will

File: contracts/registries/LicenseRegistry.sol (hunk 1 above)
Comment (kingster-will) [reply contracts/registries/LicenseRegistry.sol:205]: oh, they are in different functions.  this check is for `attachLicenseTermsToIp` and LN 235 is for `registerDerivativeIp`. they won't be invoked within the same call.

File: contracts/registries/LicenseRegistry.sol (hunk 2 above)
Comment (kingster-will) [reply contracts/registries/LicenseRegistry.sol:196]: It should be by accident. moved back.

File: test/foundry/modules/licensing/LicensingModule.t.sol
Comment (kingster-will) [reply test/foundry/modules/licensing/LicensingModule.t.sol:1]: will addresses in the next PR.

File: test/foundry/integration/e2e/e2e.t.sol
Comment (kingster-will) [reply test/foundry/integration/e2e/e2e.t.sol:1]: renamed.

**Review by jdubpark — APPROVED** [review]

**Comment by kingster-will** [comment]
Closes #39 
Closes #46
Closes #40

**Pull request merged**